File: src/main/scala/gt201101/ImperativeEuler7.scala
package gt201101

import collection.mutable.ListBuffer

/**
 * @author Ceki G&uuml;c&uuml;
*/
     
object ImperativeEuler7 {
  val primes :ListBuffer[Int] = ListBuffer(2)
  val limit = 10001;

  def initPrimeList() = {
    var i: Int = 3
    while (primes.length < limit) {
      if (isPrime(i))
        primes += i
      i += 2
    }
  }

  def isPrime(n: Int): Boolean = {
    for (p <- primes) {
      if (n % p == 0) return false
      if (p * p > n) return true
    }
    throw new IllegalStateException("failed to determine primality of [" + n + "]" + primes)
  }

  // runs in about 55 milliseconds
  def main(args: Array[String]) = {
     val s = System.currentTimeMillis
    initPrimeList
    assert (104743== primes.last)
    println("exect time "+(System.currentTimeMillis-s) + " in ms")

  }
}

File: src/main/java/gt201101/ImperativeEuler7.java
package gt201101;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Ceki G&uuml;c&uuml;
 */

class JavaEuler7 {

  List<Integer> primes = new ArrayList<Integer>();
  int limit = 10001;

  JavaEuler7() {
    primes.add(2);
  }

  void initPrimeList() {
  int i = 3;
    while (primes.size() < limit) {
      if (isPrime(i))
        primes.add(i);
      i += 2;
    }
  }

  boolean isPrime(int n) {
    for (int i = 0; i < primes.size(); i++) {
      int p = primes.get(i);
      if (n % p == 0) return false;
      if (p * p > n) return true;
    }
    throw new IllegalStateException("failed to determine primality of [" + n + "]" + primes);
  }


  // runs in about 21 milliseconds
  public static void main(String[] args) {
    JavaEuler7 je7 = new JavaEuler7();
    long start = System.currentTimeMillis();
    je7.initPrimeList();
    long end = System.currentTimeMillis();
    System.out.println("initPrimeList() operation took "+(end-start)+"milliseconds");

  }
}